Adding IEM-DCR module

jherbert-isa opened the pull request:
## Describe your changes
This pull request is about adding a new module for communicating with Isabellenhuette IEM-DCR powermeter.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements. Remark: Tests were performed with embedding the module in existing run-sil-dc.sh environment. If a fully automated test is required, please let me know.

florinmihut reviewed (changes requested):
We are not accepting different formatting rules or lint rules. 
Tabs are not allowed, etc.
Please fix this

barsnick commented on modules/IsabellenhuetteIemDcr/main/isabellenhuette_IemDcr_controller.cpp:
```diff
@@ -0,0 +1,408 @@
+	} else if(signChar == '-') {
+		nowWithOffset = std::chrono::system_clock::to_time_t(now - timeOffset);
+	} else {
+		throw std::runtime_error("manifest.yaml: Format of timezone not supported. Expected: something like \"+0100\".");
```
`std::stoi()` can also throw, e.g. if it finds only letters, so the `catch` block should be extended.

And the parsing of the config perhaps shouldn't be done on every call of the helper. I suggest that parsing to a member variable `timeOffset` (and failing/exiting on unexpected format) should happen in `init()`.

BTW, EVerest uses snake_case for variable names, CamelCase for classes and structs.

And you can use more `const` for intermediate calculations such as `now`, `signChar`, `offsetHours`, `offsetMinutes`, to help understand that the value will not be modified again (and assist the compiler in optimizing).

jherbert-isa replied:
Thanks for reviewing. Is CamelCase also to be used for method names inside classes?

If it is OK, I will omit adapting class names to CamelCase in order to avoid changing file names in this PR.

barsnick replied:
Yeah, I wonder whether there's a style guide somewhere. Do check the other modules in everest-core (though some do still make use of deprecated style).

Variables and member functions use snake_case, as do namespaces. Classes use CamelCase (sometimes camelCase, for some reason), as do types.

Don't worry about renaming files, GitHub covers that nicely, i.e. in terms of reviewing.

As @florinmihut mentioned, do remove all those extra files, and please reformat the code using [the top-level clang-format config file](https://github.com/EVerest/everest-core/blob/main/.clang-format) for readability. Some of your indentation is "wild".

Remember to sign off _all_ commits in the PR, in order to pass the DCO check.

jherbert-isa commented:
I've added some commits with the requested changes. Static code analysis now only shows "notice" findings which all seems to be false-positives. One question regarding DCO: As some commits of the past were not performed with signoff, I've added a `past_commits.txt` file as described [here](https://github.com/lf-energy/tac/blob/main/process/contribution_guidelines.md#signoff-for-commits-where-the-dco-signoff-was-missed). Shall I instead do a git rebase (e.g. `git rebase HEAD~15 --signoff`)? Although I'm the only one implementing the driver, I'm not sure if this will fit into your already started review process.

mhei commented on modules/IsabellenhuetteIemDcr/main/powermeterImpl.cpp:
```diff
@@ -0,0 +1,281 @@
+        EVLOG_error << "Incorrect module config: parameter ip_address is empty." << std::endl;
+        throw std::runtime_error("ip_address invalid. Please check configuration.");
+    }
+    if (mod->config.port_http < 0) {
```
You can here use the comfort of the manifests to specify minimum and maximum values. Then you don't need to check this stuff here.

mhei commented on modules/IsabellenhuetteIemDcr/manifest.yaml:
```diff
@@ -0,0 +1,58 @@
+    default: "192.168.60.12"
+  port_http:
+    description: HTTP-Port of the power meter API.
+    type: integer
```
As mentioned above, just add
```
    minimum: 1
    maximum: 65535
```
so that EVerest bails out if user configures stuff not in this range.

mhei commented on modules/IsabellenhuetteIemDcr/manifest.yaml:
```diff
@@ -0,0 +1,58 @@
+  timezone:
+    description: The timezone offset information according to ISO8601 (version without colon).
+    type: string
+    default: "+0100"
```
This looks conceptual wrong to me. While the timezone as-that is bound to the location of the charger (and these timezones are usually referred to via strings, e.g. "Europe/Berlin"), the string you used here must  actually change between daylight saving periods. You don't want users to reconfigure the station 2 times per year, do you?

I did not yet understand from the remaining code, whether the timezone can be changed for each charging session or whether this needs a reboot of the meter?

But in any case, I strongly recommend to determine this value on-the-fly from the underlying Linux system at runtime. You can/should expect that users running EVerest should provide a proper configured environment.

jherbert-isa replied:
Thanks for bringing this up. I've added a feature for automatic handling of DST in the new commit.

mhei commented on modules/IsabellenhuetteIemDcr/manifest.yaml:
```diff
@@ -0,0 +1,58 @@
+metadata:
+  license: https://opensource.org/licenses/Apache-2.0
+  authors:
+    - Josef Herbert, josef.herbert@isabellenhuette.com
```
I'd recommend the typical e-mail address format: Your Name <your.name@...>
